### PR TITLE
Chat status tracking

### LINKED ISSUE #54: Automatically disconnect when other party disconnects
When one member of a chat disconnects, the other user is stuck in a limbo where they do not know that the chat has ended, and messages are sent to a null user id. 

Experimentally a solution with an extra call, called once every x get messages calls, which returned the status of the active chat works, and a route was created. However this solution was buggy, and another solution may be cleaner to implement

### PR DESCRIPTION
**Pull request by arades79**
### Summary

This PR changes the collect messages endpoint to return json that includes chat status and incoming messages.

This allows the server to communicate when the other party of the chat has indicated that counseling is complete.

The client code was modified to decode the json and look for the chat status to disconnect when the status indicates the other party has left.

This means when one member of a chat disconnects, the client of the other participant will automatically disconnect.

This PR fixes issue #54 
### Checklist

- [x] If code changes were made then they have been tested.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: hyperdome/client/threads.py [hunk 1]
```diff
@@ -225,14 +225,19 @@ def __init__(self, session: requests.Session, server: Server, uid: str):
     @QtCore.pyqtSlot()
     def run(self):
         try:
-            new_messages = get_messages(self.server, self.session, self.uid)
-            self.signals.success.emit(new_messages)
+            message_response = get_messages(self.server, self.session, self.uid)
+            if message_response["chat_status"] == "CHAT_ACTIVE":
+                self.signals.success.emit(message_response["messages"])
+            elif message_response["chat_status"] == "CHAT_OVER":
+                self.signals.error.emit("chat ended")
```
Comment (stevenpitts) [inline hyperdome/client/threads.py:232]: Could be good use of ternary but w/e

File: hyperdome/client/threads.py [hunk 2]
```diff
@@ -225,14 +225,19 @@ def __init__(self, session: requests.Session, server: Server, uid: str):
...
         except requests.HTTPError:
             self.signals.error.emit("Counselor not in chat")
         except requests.RequestException:
             self.signals.error.emit("Error in get messages request")
         except MethodNotAllowed:
             self.signals.error.emit("not allowed")
+        except KeyError:
+            self.signals.error.emit("API error")
```
Comment (stevenpitts) [inline hyperdome/client/threads.py:240]: I hate swallowed exceptions
At least log it ;~;

File: hyperdome/common/onion.py [hunk 3]
```diff
@@ -287,9 +287,7 @@ def connect(
                             self.obfs4proxy_file_path
                         )
                     )
-                    with open(
-                        get_resource_path("torrc_template-obfs4")
-                    ) as o:
+                    with open(get_resource_path("torrc_template-obfs4")) as o:
```
Comment (stevenpitts) [inline hyperdome/common/onion.py:290]: I know this is just reformatted code,  but I'm not a fan of any one-letter variable other than `i`

File: hyperdome/server/web.py [hunk 4]
```diff
@@ -182,7 +185,10 @@ def counseling_complete():
             other_user = self.active_chat_user_map[sid]
             self.active_chat_user_map.pop(sid)
             self.pending_messages.pop(sid, "")
-            self.active_chat_user_map.update({other_user: ""})
+            try:
+                self.active_chat_user_map[other_user] = ""
+            except KeyError:
+                pass
```
Comment (stevenpitts) [inline hyperdome/server/web.py:191]: You might wanna explain what it means if `other_user` isn't there

**Review by stevenpitts — APPROVED** [review]

File: hyperdome/server/web.py (hunk 4 above)
Comment (arades79) [reply hyperdome/server/web.py:191]: it means that the other user has already disconnected, could be more clear

File: hyperdome/common/onion.py (hunk 3 above)
Comment (arades79) [reply hyperdome/common/onion.py:290]: how would you feel if it was `with open() as f`? Agree that `o` is not descriptive at all

Comment (stevenpitts) [reply hyperdome/common/onion.py:290]: It's not violating any standards, so you don't have to change it.
(Fuck you if you make a variable l or I though)
I always do in_file and out_file for this kinda stuff, but it's up to you

File: hyperdome/client/threads.py (hunk 2 above)
Comment (arades79) [reply hyperdome/client/threads.py:240]: soon™️ 
it's only here so I know what exceptions we're catching when I rewrite this whole thing and implement logging.

The current logging system is literally a wrapped print statement which I find unacceptable.

**Pull request merged**